service/storage_service.cc

```diff
@@ -763,18 +784,48 @@ void storage_service::join_token_ring(int delay) {
         MigrationManager.announceNewKeyspace(TraceKeyspace.definition(), 0, false);
 #endif
 
-    // At this point our local tokens are chosen (_bootstrap_tokens) and will not be changed unless we bootstrap again.
-    db::system_keyspace::set_bootstrap_state(db::system_keyspace::bootstrap_state::COMPLETED).get();
-
     slogger.debug("Setting tokens to {}", _bootstrap_tokens);
     // This node must know about its chosen tokens before other nodes do
     // since they may start sending writes to this node after it gossips status = NORMAL.
     // Therefore, in case we haven't updated _token_metadata with our tokens yet, do it now.
     _token_metadata.update_normal_tokens(_bootstrap_tokens, get_broadcast_address());
 
+    if (!db::system_keyspace::bootstrap_complete()) {
+        // If we're not bootstrapping, then we shouldn't have chosen a CDC streams timestamp yet.
+        assert(should_bootstrap() || !_cdc_streams_ts);
+    }
+
+    if (!_cdc_streams_ts && db().local().get_config().check_experimental(db::experimental_features_t::CDC)) {
+        // If we didn't choose a CDC streams timestamp at this point, then either
+        // 1. we already bootstrapped, but are upgrading from a non-CDC version (or previous version of CDC),
+        // 2. or we didn't, but we're skipping the streaming phase (seed node/auto_bootstrap=off)
+        //    and directly joining the token ring.
+
+        // In the first case (updating), we'll propose the first generation
+        // if we conclude that we're the node responsible for this.
+        // In the second case we'll always propose, just as normally bootstrapping nodes do.
+
+        if (!db::system_keyspace::bootstrap_complete() /* second case */
+                || cdc::should_propose_first_generation(get_broadcast_address(), _gossiper)) {
+
+            _cdc_streams_ts = cdc::make_new_cdc_generation(
+                    _bootstrap_tokens, _token_metadata, _sys_dist_ks.local(), get_ring_delay(), _for_testing);
+        }
+    }
+
+    // Persist the CDC streams timestamp before we persist bootstrap_state = COMPLETED.
+    if (_cdc_streams_ts) {
```

Comment:
Perhaps we can keep state in a new class cdc::controller, which becomes a member of storage_service. So stroage_service doesn't need to understand cdc stream timestamps.

Reply:
`storage_service` must understand CDC stream timestamps so it can add them to gossiper atomically together with tokens and status. E.g. `set_gossip_tokens`:
```
    _gossiper.add_local_application_state({
        { gms::application_state::TOKENS, value_factory.tokens(tokens) },
        { gms::application_state::CDC_STREAMS_TIMESTAMP, value_factory.cdc_streams_timestamp(cdc_streams_ts) },
        { gms::application_state::STATUS, value_factory.normal(tokens) }
    }).get();
```

In the future, when we move away from gossiper-based CDC stream timestamp management, I'll be able to remove this stuff from `storage_service` (and hopefully it won't have to be replaced with some other stuff).